pkg/controller/csr/csr_controller.go

```diff
@@ -0,0 +1,223 @@
+					{
+						Type:    certificatesv1.CertificateDenied,
+						Message: err.Error(),
+						Reason:  err.Error(),
+						Status:  corev1.ConditionTrue,
+					},
+				}
+				log.Error(err, "Rejecting the CSR.")
+				_, err = r.certificatesClient.CertificateSigningRequests().UpdateStatus(ctx, &csr, metav1.UpdateOptions{})
+				if err != nil {
+					return reconcile.Result{}, err
+				}
+				break
+			}
+			csr.Status.Conditions = []certificatesv1.CertificateSigningRequestCondition{
+				{
+					Type:    certificatesv1.CertificateApproved,
+					Message: "Approved",
+					Reason:  "Approved",
+					Status:  corev1.ConditionTrue,
+				},
+			}
+			reqs := r.certificatesClient.CertificateSigningRequests()
+			updatedCSR, err := reqs.UpdateApproval(ctx, csr.Name, &csr, metav1.UpdateOptions{})
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			log.Info("Approved CSR with name : %v.", csr.Name)
+
+			certificatePEM, err := certificateManager.SignCertificate(certificateTemplate, certificateTemplate.PublicKey)
+			if err != nil {
+				log.Error(err, "error signing certificate request")
+				return reconcile.Result{}, err
+			}
+			updatedCSR.Status.Certificate = certificatePEM
+
+			_, err = r.certificatesClient.CertificateSigningRequests().UpdateStatus(ctx, updatedCSR, metav1.UpdateOptions{})
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+			log.Info("Signed CSR with name : %v.", csr.Name)
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func validate(csr *certificatesv1.CertificateSigningRequest) (*x509.Certificate, error) {
+	for _, condition := range csr.Status.Conditions {
+		if (condition.Type == certificatesv1.CertificateDenied || condition.Type == certificatesv1.CertificateFailed) && condition.Status == corev1.ConditionTrue {
+			// These request statuses are deterministic/non-recoverable.
+			log.Info("will skip CSR, because its conditions won't allow us to further consider it", "status", condition.Type, "csr.name", csr.Name)
+			return nil, nil
+		}
+	}
+	firstBlock, restBlocks := pem.Decode(csr.Spec.Request)
+	if firstBlock == nil {
+		return nil, fmt.Errorf("cannot parse certificate request for CSR with name %s", csr.Name)
+	}
+	if len(restBlocks) != 0 {
+		return nil, fmt.Errorf("unexpected (multiple) pem blocks for CSR with name %s", csr.Name)
+	}
+
+	cr, err := x509.ParseCertificateRequest(firstBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+	if err = cr.CheckSignature(); err != nil {
+		return nil, err
+	}
+
+	if !certificatemanagement.TLSAssets.ValidRequestor(csr.Name, csr.Spec.Username) {
+		return nil, fmt.Errorf("invalid requestor %s for CSR with name %s", csr.Spec.Username, csr.Name)
+	}
+	if !certificatemanagement.TLSAssets.ValidDNSNames(csr.Name, cr) {
+		return nil, fmt.Errorf("invalid dns names found in CSR with name %s", csr.Name)
+	}
+	// todo: validate the ip address by using the podname from the csr and make a GET to the k8s pods api.
```

Comment:
Do we already have permissions to list all pods? Do we really need that?

Reply:
The operator currently has these permissions.

If we do not validate the IP, then technically an attacker that obtains one of our service account tokens can issue a cert for any IPSAN it wants. I don't think there is anything other than prometheus that would be making IP based requests, so the attack has very limited potential.